ir/shared/src/main/scala/org/scalajs/ir/Types.scala

```diff
@@ -62,19 +62,24 @@ object Types {
     }
   }
 
-  sealed abstract class PrimTypeWithRef extends PrimType {
-    def primRef: PrimRef = this match {
-      case VoidType    => VoidRef
-      case BooleanType => BooleanRef
-      case CharType    => CharRef
-      case ByteType    => ByteRef
-      case ShortType   => ShortRef
-      case IntType     => IntRef
-      case LongType    => LongRef
-      case FloatType   => FloatRef
-      case DoubleType  => DoubleRef
-      case NullType    => NullRef
-      case NothingType => NothingRef
+  sealed abstract class PrimTypeWithRef(private[Types] val charCode: Char) extends PrimType {
+    /* The `charCode` arguably belongs to the corresponding `PrimRef` more than
+     * to this class. We define it this way so that we don't have an apparent
+     * circular dependency between the `PrimTypeWithRef`s and their `PrimRef`s.
+     */
+
+    def primRef: PrimRef = (charCode: @switch) match {
```

Comment:
Have you considered `val primRef = PrimeRef(this)` and then get the refs from the types for the aliases below?